### ISSUE TITLE
Move expense between Events, Projects, and the parent Collective

### ISSUE THREAD

**Issue by alanna** [opencollective/opencollective#3981]
## User story 
As admin of a Collective with an Event, I want to be able to move expenses from the Event to the Collective if they are submitted to the wrong one.

## Best solution for this problem
An action visible on expenses for an Event `Move to main Collective`

**Comment by BenJam** [comment]
I think there's a more general story here about transferring money between 'collectives' (actual collectives, events and now projects) that needs to be better supported.

**Comment by alanna** [comment]
Now that Projects has been released, users are asking to be able to move expenses between Projects and the main budget.

**Comment by BenJam** [comment]
I think @SudharakaP has been looking at this no?

**Comment by SudharakaP** [comment]
@BenJam : Yes, if you are referring to Alanna's comment; 

> Now that Projects has been released, users are asking to be able to move expenses between Projects and the main budget

I think we have some support for this at, https://github.com/opencollective/opencollective/issues/4468 to improve the fund transfers between events/projects but not specifically this issue. My understanding is this issue is to tackle transferring the funds on an expense to expense basis. This for this one, I think we need to wait for design. 🤔 

cc: @Betree 

.

**Comment by Betree** [comment]
Allowing to move expenses that haven't been paid yet is not super complex. Allowing to move expenses that **have been paid** is more complex:
- We need to update the transactions in the ledger and the activities
- We need to handle the case where the parent doesn't have enough budget for this expense (by automatically moving funds between the child and the parent first?)

Overall I agree with @SudharakaP, having input from design on this one first would be great.

**Comment by alanna** [comment]
How about we start with moving unpaid expenses only? That would still be
useful.

**Comment by ryul95** [comment]
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by mlncn** [comment]
Bumping this as particularly important for collectives with credit/debit cards through their fiscal hosts.  It seems that expenses to the card can only go through the parent collective.  Can projects use the parent collective card?  Is it feasible for them to get their own cards in every case?

**Comment by alanna** [comment]
#4798 has been done and solves @mlncn 's issue, I believe (provided the card was not assigned to the wrong profile in error). The other aspects remain.

**Comment by alanna** [comment]
bump, this is coming up regularly

**Comment by Betree** [comment]
As a quick hack for this one, it has been suggested somewhere that we could start with a root action.

**Comment by alanna** [comment]
Root action would be great

**Comment by Betree** [comment]
Thanks to the contribution of @hdiniz, the root action for this is mostly done (but not deployed yet) and will provide the basis to implement it for everyone. What needs to be done before exposing this publicly:
- [ ] We need a design
- [ ] We need to update the permissions to allow other users to move expenses (only from/to collectives they administer)
- [ ] Redirect from the old URL: if there's a `COLLECTIVE_EXPENSE_MOVED` activity in the history and we're trying to visit `/${previousCollective}/expense/${id}`, it should redirect to `/${newCollective}/expense/${id}`
- [ ] We need to update the code to support moving paid expenses (aka. expenses with transactions)
- [ ] In the expense activity thread, the previous collective should be displayed as a link but it's not safe to do it unless we implement https://github.com/opencollective/opencollective/issues/5808 first

**Comment by iamronen** [comment]
reposting from Slack:
I can understand the need for such a feature, however I believe that such funding/expense related capabilities (I'm guessing there are others like it in the root actions) implemented this way, create a fundamental trust and reliability issue:

1. As the person who submitted the expense I may be wondering WTF just happened ... where did my expense go and assuming I figure out how it got there, I am wondering who has the power to do that? In order for this operation to be handled well, the submitter should be notified that this happened - so that when he can understand why he got paid from an account he did not ask to be paid from.
2. In a system that provides transparency and audit-ability, sensitive information like this should not be able to MOVE. If an operation like this has to be performed a move should be implemented as:
    - Mark the original transaction/expense as canceled and indicate why this happened
    - Create a new copy that is in the right context, indicating that it was created through bypass mechanism (because in this case the submitter did not submit this expense to the alternative account).
    - Link the canceled transaction to the new one and the new one to the old.
    - Both data-instances should exist as independent entities with both internal and referential integrity.

Any "magical" intervention like this that is not clearly documented is a seed that can grow into mistrust.

It can be useful to think about this in web3 terms: in a decentralized environment you simply don't have the privilege of having a centralized database that ANY SINGLE HUMAN BEING can manipulate.  I believe this is the kind of resilience our accounting information architecture needs to have.

Addition:
@Betree you responded: 

> this is something we already to today

In the demo the only trace I saw of this behavior is that the new expense (assuming it is, as you seem to imply, it is a copy) indicated that it was moved in the target account. It "disappeared" from the source account which, I believe, should not be possible. No one should be able to "move" an expense I submitted. Someone may be able to "help" me make a correction, by making a copy on my behalf ... but that also requires that someone communicate this to me.

Also you did not indicate that there was any validation going on: that the choice of the source account limited the choice of the target account. Is there such validation?

**Comment by Betree** [comment]
> Also you did not indicate that there was any validation going on: that the choice of the source account limited the choice of the target account. Is there such validation?

@iamronen As mentioned, this tool is only for root users (aka the core team) at the moment to help them with a very common support task. We trust people in the support team to properly communicate this change with users and verify beforehand that there's a legit use case, which is why we haven't implemented advanced verifications (_apart from the validations on really problematic stuff, like you can't move an expense to a user profile_).

As mentioned in this issue, we'll want to limit this feature to only move expenses under the same collective when exposed to everyone.

>  I believe this is the kind of resilience our accounting information architecture needs to have.

We don't support moving expenses with attached transactions, so there's accounting in play here. I'm all in for immutable ledgers, that's actually how ours is designed today. But here we're talking about moving an expense request, no money's actually involved at this point. Even when we'll release this for everyone, we may very well stay on our "Can't move paid expenses" position, as 90% of the cases we want to solve here are people who just want to move an expense right after it was submitted to address it to the right event/project. Being able to move paid expenses does not appear so useful at this point.

> Any "magical" intervention like this that is not clearly documented is a seed that can grow into mistrust.

I don't understand where's the trust issue. If I submitted my expense to `Babel` instead of `Babel > Event 1` and the admin moves it, I won't feel betrayed. Events/Projects have the same admins as their parents, so there's no change in privacy. The budget that you're going to use to pay me doesn't matter much.

When we'll expose this feature for everyone, we'll indeed need to think about notifications, redirection (as proposed in https://github.com/opencollective/opencollective/issues/3981#issuecomment-1202365331), and overall proper UX.

**Comment by alanna** [comment]
I agree with @Betree - these actions are pretty much always taken by support at the request of the user. That said, @iamronen's comments make me think we should include this kind of thing on the Activity Log #5195 for transparency and good recordkeeping.

**Comment by Betree** [comment]
> I agree with @Betree - these actions are pretty much always taken by support at the request of the user. That said, @iamronen's comments make me think we should include this kind of thing on the Activity Log #5195 for transparency and good recordkeeping.

@alanna I've already created an Activity for it, in the demo posted on Slack you'll notice that it appears on the expense page as `Expense was moved from {previousCollective}`. Because there's an Activity for it, it will appear in the Activity Log out of the box :slightly_smiling_face: 

![image](https://user-images.githubusercontent.com/1556356/182538348-c3b42f53-230d-472c-b93c-18bcc05151e2.png)

**Comment by alanna** [comment]
oh perfect thanks!

**Comment by Betree** [comment]
Re-opening as we only introduced the root action while the issue was initially about exposing this feature to all users
